include/alpaka/atomic/Op.hpp

```diff
@@ -24,17 +26,28 @@ namespace alpaka
         template<typename T>
         ALPAKA_FN_HOST_ACC auto operator()(T* const addr, T const& value) const -> T
         {
-            auto const old = *addr;
-            auto& ref = *addr;
+            static constexpr bool IsLockFree
+                = std::is_trivially_copyable_v<T> && boost::atomic_ref<T>::required_alignment <= alignof(T)
+                && boost::atomic_ref<T>::is_always_lock_free;
+            if constexpr(IsLockFree)
+            {
+                boost::atomic_ref<T> ref(*addr);
+                return ref.fetch_add(value);
+            }
+            else
+            {
+                auto const old = *addr;
+                auto& ref = *addr;
 #if BOOST_COMP_GNUC
 #    pragma GCC diagnostic push
 #    pragma GCC diagnostic ignored "-Wconversion"
 #endif
-            ref += value;
-            return old;
+                ref += value;
+                return old;
 #if BOOST_COMP_GNUC
 #    pragma GCC diagnostic pop
 #endif
+            }
```

Comment:
So this means now that if the serial backend is enabled, it will still use atomic operations, which it did not before, which is a potential pessimization. This might need some discussion, or a general architectural refactoring. It is weird to me that the `AtomicStdLibLock` implementation uses `operator()` of the passes operators, whereas the CUDA/HIP implementation just uses the passed operators for overload resolution but then provides an independent implementation. I would rather have the second architecture for all backends. But I am open for suggestions.

Reply:
Looking through what kind of atomics are used at which level, I am surprised to find that even the serial backend uses the lock based atomic implementation for the grid atomics.

Reply:
These implementations should be specializations for ´AtomicStdLibLock` or better for a new atomic backend `AtomicAtomicRef`. I think this file should not be modified.

Reply:
> Looking through what kind of atomics are used at which level, I am surprised to find that even the serial backend uses the lock based atomic implementation for the grid atomics.

This is required because you can have two kernel running in different async queues. In that case you must guarantee that the operations are atomic. Grid means that operations are atomic between grids.

Reply:
> Looking through what kind of atomics are used at which level, I am surprised to find that even the serial backend uses the lock based atomic implementation for the grid atomics.

I understood that "grid atomics" means "atomic with respect to operations running in different grids".
Then, yes, it makes sense that the serial backend uses some atomic operations for them, if the "kernel" asks for it - because there can be multiple "grids" running concurrently in multiple host threads.

Reply:
@jkelling wrote:
> I think this file should not be modified.

Yes, you are right, the changes should not be in `Op.hpp`.

Reply:
> This is required because you can have two kernel running in different async queues. In that case you must guarantee that the operations are atomic. Grid means that operations are atomic between grids.

Thank you, that makes perfect sense!